File: java/src/main/java/io/fluentcards/FactSetBuilder.java
package io.fluentcards;

import java.util.*;

public class FactSetBuilder {
    private final LinkedHashMap<String, Object> data;

    public FactSetBuilder() {
        data = new LinkedHashMap<>();
        data.put("type", "FactSet");
        data.put("facts", new ArrayList<>());
    }

    public FactSetBuilder withId(String id) {
        data.put("id", id);
        return this;
    }

    public FactSetBuilder withSpacing(Spacing spacing) {
        data.put("spacing", spacing.getValue());
        return this;
    }

    @SuppressWarnings("unchecked")
    public FactSetBuilder addFact(String title, String value) {
        LinkedHashMap<String, Object> fact = new LinkedHashMap<>();
        fact.put("title", title);
        fact.put("value", value);
        ((List<Object>) data.get("facts")).add(fact);
        return this;
    }

    @SuppressWarnings("unchecked")
    public FactSetBuilder addFactMap(Map<String, Object> fact) {
        ((List<Object>) data.get("facts")).add(fact);
        return this;
    }

    public Map<String, Object> build() {
        return data;
    }
}

File: node/fluent-cards/src/builders/FactSetBuilder.ts
import { Spacing } from '../enums.js';
import type { FactSet, Fact } from '../models.js';

/** Fluent builder for {@link FactSet} elements. */
export class FactSetBuilder {
  private readonly factSet: FactSet = { type: 'FactSet', facts: [] };

  /** Sets the unique identifier. @param id The unique identifier. @returns The builder instance for method chaining. */
  withId(id: string): this { this.factSet.id = id; return this; }
  /** Sets the spacing above the element. @param spacing The spacing value. @returns The builder instance for method chaining. */
  withSpacing(spacing: Spacing): this { this.factSet.spacing = spacing; return this; }

  /** Adds a fact by title and value, or a pre-built Fact object. @param titleOrFact The title string or a pre-built Fact. @param value The value string (required when title is a string). @returns The builder instance for method chaining. */
  addFact(title: string, value: string): this;
  addFact(fact: Fact): this;
  addFact(titleOrFact: string | Fact, value?: string): this {
    if (typeof titleOrFact === 'string') {
      this.factSet.facts!.push({ title: titleOrFact, value: value! });
    } else {
      this.factSet.facts!.push(titleOrFact);
    }
    return this;
  }

  /** Builds and returns the configured FactSet. @returns The configured FactSet instance. */
  build(): FactSet { return this.factSet; }
}
